src/transforms/regex_parser.rs

```diff
@@ -56,14 +56,78 @@ impl TransformConfig for RegexParserConfig {
 
 pub struct RegexParser {
     regexset: RegexSet,
-    patterns: Vec<Regex>,
+    patterns: Vec<CompiledRegex>, // indexes correspend to RegexSet
     field: Atom,
     drop_field: bool,
     drop_failed: bool,
     target_field: Option<Atom>,
     overwrite_target: bool,
+}
+
+struct CompiledRegex {
+    regex: Regex,
     capture_names: Vec<(usize, Atom, Conversion)>,
-    capture_locs: Vec<CaptureLocations>,
+    capture_locs: CaptureLocations,
+}
+
+impl CompiledRegex {
+    fn new(regex: Regex, types: &HashMap<Atom, Conversion>) -> CompiledRegex {
+        // Calculate the location (index into the capture locations) of
+        // each named capture, and the required type coercion.
+        let capture_names: Vec<(usize, Atom, Conversion)> = regex
+            .capture_names()
+            .enumerate()
+            .filter_map(|(idx, cn)| {
+                cn.map(|cn| {
+                    let cn: Atom = cn.into();
+                    let conv = types.get(&cn).unwrap_or(&Conversion::Bytes);
+                    (idx, cn, conv.clone())
+                })
+            })
+            .collect();
+        let capture_locs = regex.capture_locations();
+        CompiledRegex {
+            regex,
+            capture_names,
+            capture_locs,
+        }
+    }
+
+    /// Returns a list of captures (name, value) or None if the regex does not
+    /// match
+    fn captures<'a>(&'a mut self, value: &[u8]) -> Option<Vec<(&'a Atom, Value)>> {
+        match self.regex.captures_read(&mut self.capture_locs, value) {
+            Some(_) => {
+                let values = self
+                    .capture_names
+                    .iter()
+                    .filter_map(
+                        |(idx, name, conversion)| match self.capture_locs.get(*idx) {
+                            Some((start, end)) => {
+                                let capture: Value = value[start..end].into();
+
+                                match conversion.convert(capture) {
+                                    Ok(value) => Some((name, value)),
+                                    Err(error) => {
+                                        debug!(
+                                            message = "Could not convert types.",
+                                            name = &name[..],
+                                            %error,
+                                            rate_limit_secs = 30
+                                        );
+                                        None
+                                    }
+                                }
+                            }
+                            None => None,
```

Comment:
Just to clarify, this would a `map_or` yes? Something like `self.capture_locs.get(*idx).map_or(None, |(start,end)| ...)` right? Otherwise it wraps it in another `Option` and I get:

```
value of type `std::vec::Vec<(&string_cache::atom::Atom<string_cache::atom::EmptyStaticAtomSet>, event::Value)>` cannot be built from `std::iter::Iterator<Item=std::option::Option<(&string_cache::atom::Atom<string_cache::atom::EmptyStaticAtomSet>, event::Value)>>`
```

Reply:
I misspoke, because I missed the inner match resulted in an option too, so it would be `.and_then` instead of `.map`.